threadpoolctl.py

```diff
@@ -60,56 +67,322 @@ class _dl_phdr_info(ctypes.Structure):
     _RTLD_NOLOAD = ctypes.DEFAULT_MODE
 
 
-# List of the supported libraries. The items are indexed by the name of the
-# class to instantiate to create the library controller objects. The items hold
-# the possible prefixes of loaded shared objects, the name of the internal_api
-# to call, the name of the user_api and potentially some symbols that the library is
-# expected to have (this is necessary to distinguish between the blas implementations
-# when they are all renamed "libblas.dll" on conda-forge on windows).
-_SUPPORTED_LIBRARIES = {
-    "OpenMPController": {
-        "user_api": "openmp",
-        "internal_api": "openmp",
-        "filename_prefixes": ("libiomp", "libgomp", "libomp", "vcomp"),
-    },
-    "OpenBLASController": {
-        "user_api": "blas",
-        "internal_api": "openblas",
-        "filename_prefixes": ("libopenblas", "libblas"),
-        "check_symbols": ("openblas_get_num_threads", "openblas_get_num_threads64_"),
-    },
-    "MKLController": {
-        "user_api": "blas",
-        "internal_api": "mkl",
-        "filename_prefixes": ("libmkl_rt", "mkl_rt", "libblas"),
-        "check_symbols": ("MKL_Get_Max_Threads",),
-    },
-    "BLISController": {
-        "user_api": "blas",
-        "internal_api": "blis",
-        "filename_prefixes": ("libblis", "libblas"),
-        "check_symbols": ("bli_thread_get_num_threads",),
-    },
-}
+class LibController(ABC):
+    """Abstract base class for the individual library controllers
+
+    A library controller must expose the following class attributes:
+        - user_api : str
+            Usually the name of the library or generic specification the library
+            implements.
+        - internal_api : str
+            Usually the name of the library, or concrete implementation of some
+            specification.
+        - filename_prefixes : tuple
+            Possible prefixes of the shared library's filename that allow to
+            identify the library. e.g. "libopenblas" for libopenblas.so.
+
+    A library contoller must implement the following methods: `get_num_threads`,
+    `set_num_threads` and `get_version`.
+
+    The following information will be exposed in the info dict:
+      - user_api : user API.
+      - internal_api : internal API.
+      - num_threads : the current thread limit.
+      - prefix : prefix of the shared library's filename.
+      - filepath : path to the loaded shared library.
+      - version : version of the library (if available).
+
+    In addition, each library controller may expose internal API specific entries. They
+    must be set as attributes in the `set_additional_attributes method`.
+    """
```

Comment:
Maybe something missing in the doc is to explain that we will loop through the linked libs to find libs matching the `filename_prefix` and that these libs will be available as `_dynlib`?
Otherwise, you need to know the internals of the API pretty well to be able to implement this

Reply:
I was actually in the middle of making it public and documenting it :) I just pushed that. Let me know if you find it clear enough

Reply:
> I just pushed that. Let me know if you find it clear enough

where is this? 479b111 does not explain that the filename of dynamic libraries is used to match the library and that threadpoolctl inspect this filename via a look-up to the C runtime (via `dl_iterate_phdr` and what happens in `_make_controller_from_path`).

Reply:
I added something